flow/connectors/kafka/kafka.go

```diff
@@ -0,0 +1,361 @@
+	dialer       kafka.Dialer
+	connection   *kafka.Conn
+	writer       *kafka.Writer
+	readerConfig *kafka.ReaderConfig
+}
+
+func NewKafkaConnector(ctx context.Context,
+	kafkaProtoConfig *protos.KafkaConfig) (*KafkaConnector, error) {
+	securityProtocol := kafkaProtoConfig.SecurityProtocol
+	username := kafkaProtoConfig.Username
+	password := kafkaProtoConfig.Password
+
+	brokers := strings.Split(kafkaProtoConfig.Servers, ",")
+
+	dialer := &kafka.Dialer{
+		Timeout:   10 * time.Second,
+		DualStack: true,
+	}
+
+	writer := &kafka.Writer{
+		Addr:                   kafka.TCP(brokers...),
+		AllowAutoTopicCreation: true,
+	}
+
+	if securityProtocol == "SASL_SSL" {
+		saslMechanism := plain.Mechanism{
+			Username: username,
+			Password: password,
+		}
+		caCertPool := x509.NewCertPool()
+		caCertPool.AppendCertsFromPEM([]byte(kafkaProtoConfig.SslCertificate))
+		tlsConfig := &tls.Config{
+			RootCAs: caCertPool,
+		}
+		dialer.TLS = tlsConfig
+		dialer.SASLMechanism = saslMechanism
+		writer.Transport = &kafka.Transport{
+			TLS:  tlsConfig,
+			SASL: saslMechanism,
+		}
+	}
+	conn, err := dialer.Dial("tcp", brokers[0])
+	if err != nil {
+		return nil, fmt.Errorf("failed to open connection to Kafka cluster: %w", err)
+	}
+	controller, err := conn.Controller()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get controller to leader of Kafka cluster: %w", err)
+	}
+	leaderConn := net.JoinHostPort(controller.Host, strconv.Itoa(controller.Port))
+	controllerConn, err := dialer.Dial("tcp", leaderConn)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get connection to leader controller: %w", err)
+	}
+	log.Println("Obtained controller connection.")
+	writer.Addr = controllerConn.RemoteAddr()
+	readerConfig := &kafka.ReaderConfig{
+		Brokers: append(brokers, leaderConn),
+		Dialer:  dialer,
+	}
+	return &KafkaConnector{
+		ctx:          ctx,
+		dialer:       *dialer,
+		connection:   controllerConn,
+		writer:       writer,
+		readerConfig: readerConfig,
+	}, nil
+}
+
+func (c *KafkaConnector) Close() error {
+	if c == nil || c.connection == nil {
+		return nil
+	}
+	err := c.connection.Close()
+	if err != nil {
+		return fmt.Errorf("error while closing connection to Kafka peer: %w", err)
+	}
+	return nil
+}
+
+func (c *KafkaConnector) ConnectionActive() bool {
+	if c == nil || c.connection == nil {
+		return false
+	}
+	_, err := c.connection.ReadPartitions()
+	return err != nil
+}
+
+func (c *KafkaConnector) NeedsSetupMetadataTables() bool {
+	return true
+}
+
+func (c *KafkaConnector) SetupMetadataTables(jobName string) error {
+	metadataTopicName := "peerdb_" + jobName
+	createErr := c.connection.CreateTopics(kafka.TopicConfig{
+		Topic:             metadataTopicName,
+		NumPartitions:     1,
+		ReplicationFactor: 1,
+	})
+
+	if createErr != nil {
+		return fmt.
+			Errorf("failed to create Kafka topic: %w", createErr)
+	}
+	return nil
+}
+
+func (c *KafkaConnector) GetLastOffset(jobName string) (*protos.LastSyncState, error) {
```

Comment:
add a couple of comments, 1 near the place where we set num partitions for this topic to 1 as well.